content/blog/2019-11-15-profile-guided-basic-block-reordering.md

```diff
@@ -0,0 +1,94 @@
+
+15:                                               ; preds = %2
+  store i32 6, i32* %7, align 4
+  br label %16
+
+16:                                               ; preds = %15, %14
+  ...
+  ret i32 %19
+}
+
+```
+This simply put the blocks in the original order. If we know that the input `a` is almost always smaller than 5, then we could move 15 and 16 block before 14, such that the instruction cache would not need to fetch 14. In the case where 14 branch has many more instructions, fetching 14 instead of 15 can lead to an instruction cache miss.
+
+In this project, we used the [built-in profile generation via the instrumentation tool in Clang](https://clang.llvm.org/docs/UsersManual.html#profiling-with-instrumentation) to collect profiles.
+
+## Block Reordering Algorithm
+
+We implemented one variant of algo2 in Profile Guided Code Positioning as one LLVM pass. The frequencies of each branch in conditional jump and switch instrucitons are used as edges weight. We also experimented with marking the weight of unconditional branches to be infinity.
+
+The algorithm first identifies frequently executed "chains", i.e., paths of basic blocks. It starts by setting all basic blocks as individual chains. Then the edges are iterated from highest weight to lowest. For each edge, if it is connecting tail and head of two chains, the chains will be merged. We implemented this by maintaining the start of chains and the next block relationships in hashmaps. Then we iterate through all the block terminators to collect edges with weights and rank them by weights. Finally we merge the chains if and only if the edge is jumping from the end of a chain to the start of a chain.
+
+After the chains are identified, they are arranged sequentially based on weights over edges cross chains. In our implementation, we decide the next chain to put in the way that the current chain has one branch with largest weight point to the start of it.
+
+
+## Evaluation
+
+### Initial Testing
+Initially we tested our pass on simple programs with only one conditional branch, where one branch is more frequently executed than the other. Our pass successfully rearranged the more used pass to directly after the block before branching. We also tested on more complicated branching scenarios.
+
+### Benchmark Testing
+
+We also evaluated the pass on [PARSEC](https://parsec.cs.princeton.edu/). Due to machine and time constraints, we only managed to run streamcluster programs with profiles provided by running the simulated inputs. We use `parsecmgmt` to run the benchmarks, and report the branch misses as well as the total runtimes. We compared clang (no optimizations) with only our optimization pass against clang (no optimizations). Unfortunately, there isn't a significant improvement in either the branch misses or the runtime. We suspect there is at least some improvement in streamcluster, considering the drastically lowered standard deviation and modestly reduced branch misses.
```

Comment:
Can you clarify what a "branch miss" is? Do you mean a branch misprediction? If so, how did you measure the misprediction rate?

Reply:
The branch miss is output by `parsecmgmt`.  I thought branch miss was standard nomenclature, but I guess not.

Reply:
I did just realize that it really shouldn't be branch misses that we cared about, but icache misses right? I suppose it's possible that rearranging the blocks could effect branch misses (due to lower level optimization stuff), but our optimization specifically should be trying to improve icache locality.

Reply:
Yeah, agreed. There's a chance it would affect branch predictability, but it could just as easily make branches _less_ predictable while still improving performance because of I-cache locality.